sdk/internal/perf/perf.go

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+package perf
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+	"text/tabwriter"
+	"time"
+
+	"github.com/spf13/pflag"
+)
+
+var (
+	debug     bool
+	Duration  int
+	TestProxy string
+	WarmUp    int
+	Parallel  int
+	wg        sync.WaitGroup
+)
+
+type PerfTest interface {
+	// GetMetadata returns the name of the test
+	GetMetadata() PerfTestOptions
+
+	GlobalSetup(context.Context) error
+
+	Setup(context.Context) error
+
+	Run(context.Context) error
+
+	Cleanup(context.Context) error
+
+	GlobalCleanup(context.Context) error
+}
+
+type HTTPClient interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+type PerfTestOptions struct {
+	ParallelIndex int
+	ProxyInstance HTTPClient
+	Name          string
+}
+
+type NewPerfTest func(options *PerfTestOptions) PerfTest
+
+func runGlobalSetup(p PerfTest) error {
+	setRecordingMode("live")
+	err := p.GlobalSetup(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func runSetup(p PerfTest) error {
+	err := p.Setup(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// runTest takes care of the semantics of running a single iteration. It returns the number of times the test ran as an int, the exact number
+// of seconds the test ran as a float64, and any errors.
+func runTest(p PerfTest, c chan runResult) {
+	defer wg.Done()
+
+	// If we are using the test proxy need to set up the in-memory recording.
+	if TestProxy != "" {
+		// First request goes through in Live mode
+		setRecordingMode("live")
+		err := p.Run(context.Background())
+		if err != nil {
+			c <- runResult{err: err}
+		}
+
+		// 2nd request goes through in Record mode
+		setRecordingMode("record")
+		err = start(p.GetMetadata().Name, nil)
+		if err != nil {
+			panic(err)
+		}
+		err = p.Run(context.Background())
+		if err != nil {
+			c <- runResult{err: err}
+		}
+		err = stop(p.GetMetadata().Name, nil)
+		if err != nil {
+			panic(err)
+		}
+
+		// All ensuing requests go through in Playback mode
+		setRecordingMode("playback")
+		err = start(p.GetMetadata().Name, nil)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	if WarmUp > 0 {
```

Comment:
The per-second results should be printed during both warmup and duration, in separate sections of the output.  Example from .NET:

```
=== Warmup ===
    Current             Total          Average       CPU    WorkingSet    PrivateMemory
 19,516,504        19,516,504    19,451,361.00     8.39%        35.18M           60.59M
 22,341,875        41,858,379    20,682,285.66     8.70%        35.76M           60.71M
 21,822,398        63,680,777    21,272,532.56     8.32%        35.80M           60.71M

=== Results ===
Completed 63,680,777 operations in a weighted-average of 2.99s (21,272,532.56 ops/s, 0.000 s/op, 8.46% CPU)

=== Test ===
    Current             Total          Average       CPU    WorkingSet    PrivateMemory
 22,128,980        22,128,980    21,837,353.99     8.35%        35.88M           60.68M
 23,162,566        45,291,546    22,275,325.98     8.30%        35.90M           60.71M
 21,860,969        67,152,515    22,286,358.33     8.37%        35.97M           60.72M

=== Results ===
Completed 67,152,515 operations in a weighted-average of 3.01s (22,286,358.33 ops/s, 0.000 s/op, 8.33% CPU)
```

Reply:
Added